File: service/src/main/java/no/nav/melosys/service/kontroll/feature/unntak/kontroll/AnmodningUnntakKontrollsett.java
package no.nav.melosys.service.kontroll.feature.unntak.kontroll;

import java.util.Set;
import java.util.function.Function;

import no.nav.melosys.service.kontroll.feature.arbeidutland.kontroll.ArbeidUtlandKontroll;
import no.nav.melosys.service.kontroll.feature.unntak.data.AnmodningUnntakKontrollData;
import no.nav.melosys.service.validering.Kontrollfeil;

public class AnmodningUnntakKontrollsett {

    public static Set<Function<AnmodningUnntakKontrollData, Kontrollfeil>> hentRegler() {
        return REGLER_ANMODNING_UNNTAK;
    }

    private static Set<Function<AnmodningUnntakKontrollData, Kontrollfeil>> REGLER_ANMODNING_UNNTAK = Set.of(
        AnmodningUnntakKontroll::harRegistrertAdresse,
        AnmodningUnntakKontroll::anmodningsperiodeManglerSluttdato,
        AnmodningUnntakKontroll::kunEnArbeidsgiver,
        kontrollData -> ArbeidUtlandKontroll.arbeidsstedManglerFelter(kontrollData.behandlingsgrunnlagData()),
        kontrollData -> ArbeidUtlandKontroll.foretakUtlandManglerFelter(kontrollData.behandlingsgrunnlagData())
    );
}

File: service/src/main/kotlin/no/nav/melosys/service/kontroll/feature/anmodningomunntak/kontroll/AnmodningUnntakKontrollsett.kt
package no.nav.melosys.service.kontroll.feature.anmodningomunntak.kontroll

import no.nav.melosys.service.kontroll.feature.anmodningomunntak.data.AnmodningUnntakKontrollData
import no.nav.melosys.service.kontroll.feature.arbeidutland.kontroll.ArbeidUtlandKontroll.Companion.arbeidsstedLandManglerFelter
import no.nav.melosys.service.kontroll.feature.arbeidutland.kontroll.ArbeidUtlandKontroll.Companion.foretakUtlandManglerFelter
import no.nav.melosys.service.validering.Kontrollfeil
import java.util.function.Function

typealias Regelsett = Set<Function<AnmodningUnntakKontrollData, Kontrollfeil?>>

object AnmodningUnntakKontrollsett {
    fun hentRegler(): Regelsett {
        return REGLER_ANMODNING_UNNTAK
    }

    private val REGLER_ANMODNING_UNNTAK: Regelsett = setOf(
        Function { AnmodningUnntakKontroll.harRegistrertAdresse(it) },
        Function { AnmodningUnntakKontroll.anmodningsperiodeManglerSluttdato(it) },
        Function { AnmodningUnntakKontroll.kunEnArbeidsgiver(it) },
        Function { arbeidsstedLandManglerFelter(it.mottatteOpplysningerData) },
        Function { foretakUtlandManglerFelter(it.mottatteOpplysningerData) },
        Function { AnmodningUnntakKontroll.storbritanniaKonvensjonBruktForTidlig(it) }
    )
}
